### ISSUE TITLE
Question: How to change the width of a tab

### ISSUE THREAD

**Issue by liz3** [FXMisc/RichTextFX#699]
Hello,

how do i change the Width of the Tab Character?
The Solution from Issue https://github.com/FXMisc/RichTextFX/issues/86 would be okay, but does not work anymore.
Is there a new way to do this?

Regards
Yann "Liz3" HN

**Comment by JordanMartinez** [comment]
Yes, the API was changed slightly in WellBehavedFX 0.3, you would now write something like:
````java
InputMap<KeyEvent> im = InputMap.consume(
    EventPattern.keyTyped("\t"), 
    e -> area.replaceSelection("    ")
);
Nodes.addInputMap(area, im);
````

**Comment by liz3** [comment]
Worked, Thanks!

**Comment by cerebrosoft** [comment]
Using the code above, a tab key yields both a tab and 4 spaces (in that order) being added to the code editor... is there something else I need to do to make sure the tab doesn't get inserted?

**Comment by JordanMartinez** [comment]
Do you also have something for `area.setOnKeyTyped()`? If so, then that handler will be run, too.

Which version of RTFX are you using?

**Comment by cerebrosoft** [comment]
I'm on 0.8.2.  My whole class is below (no setOnKeyTyped())

```
import java.util.Collection;
import java.util.Collections;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.fxmisc.richtext.CodeArea;
import org.fxmisc.richtext.LineNumberFactory;
import org.fxmisc.richtext.model.StyleSpans;
import org.fxmisc.richtext.model.StyleSpansBuilder;
import org.fxmisc.wellbehaved.event.EventPattern;
import org.fxmisc.wellbehaved.event.InputMap;
import org.fxmisc.wellbehaved.event.Nodes;

import javafx.scene.input.KeyEvent;

public class ScriptEditorSupport {

    private static final String[] KEYWORDS = new String[] {
        "abstract", "assert", "boolean", "break", "byte",
        "case", "catch", "char", "class", "const",
        "continue", "default", "do", "double", "else",
        "enum", "extends", "final", "finally", "float",
        "for", "goto", "if", "implements", "import",
        "instanceof", "int", "interface", "long", "native",
        "new", "package", "private", "protected", "public",
        "return", "short", "static", "strictfp", "super",
        "switch", "synchronized", "this", "throw", "throws",
        "transient", "try", "void", "volatile", "while"
    };

    private static final String KEYWORD_PATTERN = "\\b(" + String.join("|", KEYWORDS) + ")\\b";
    private static final String PAREN_PATTERN = "\\(|\\)";
    private static final String BRACE_PATTERN = "\\{|\\}";
    private static final String BRACKET_PATTERN = "\\[|\\]";
    private static final String SEMICOLON_PATTERN = "\\;";
    private static final String STRING_PATTERN = "\"([^\"\\\\]|\\\\.)*\"";
    private static final String COMMENT_PATTERN = "//[^\n]*" + "|" + "/\\*(.|\\R)*?\\*/";

    private static final Pattern PATTERN = Pattern.compile(
        "(?<KEYWORD>" + KEYWORD_PATTERN + ")"
            + "|(?<PAREN>" + PAREN_PATTERN + ")"
            + "|(?<BRACE>" + BRACE_PATTERN + ")"
            + "|(?<BRACKET>" + BRACKET_PATTERN + ")"
            + "|(?<SEMICOLON>" + SEMICOLON_PATTERN + ")"
            + "|(?<STRING>" + STRING_PATTERN + ")"
            + "|(?<COMMENT>" + COMMENT_PATTERN + ")");

    private CodeArea codeArea;

    public CodeArea getComponent() {
        if (codeArea != null) return codeArea;

        codeArea = new CodeArea();
        codeArea.setParagraphGraphicFactory(LineNumberFactory.get(codeArea));

        codeArea.richChanges()
            .filter(ch -> !ch.getInserted().equals(ch.getRemoved()))
            .subscribe(change -> {
                codeArea.setStyleSpans(0, computeHighlighting(codeArea.getText()));
            });

        InputMap<KeyEvent> im = InputMap.consume(
            EventPattern.keyTyped("\t"),
            e -> codeArea.replaceSelection("    "));
        Nodes.addInputMap(codeArea, im);

        return codeArea;
    }

    public void setText(String text) {
        getComponent().replaceText(0, getComponent().getText().length(), text);
    }

    public String getText() {
        return getComponent().getText();
    }

    private static StyleSpans<Collection<String>> computeHighlighting(String text) {
        Matcher matcher = PATTERN.matcher(text);
        int lastKwEnd = 0;
        StyleSpansBuilder<Collection<String>> spansBuilder = new StyleSpansBuilder<>();
        while (matcher.find()) {
            String styleClass =
                matcher.group("KEYWORD") != null ? "keyword" : matcher.group("PAREN") != null ? "paren" : matcher.group("BRACE") != null ? "brace" : matcher.group("BRACKET") != null ? "bracket" : matcher.group("SEMICOLON") != null ? "semicolon" : matcher.group("STRING") != null ? "string" : matcher.group("COMMENT") != null ? "comment" : null;
            // never happens
            assert styleClass != null;
            spansBuilder.add(Collections.emptyList(), matcher.start() - lastKwEnd);
            spansBuilder.add(Collections.singleton(styleClass), matcher.end() - matcher.start());
            lastKwEnd = matcher.end();
        }
        spansBuilder.add(Collections.emptyList(), text.length() - lastKwEnd);
        return spansBuilder.create();
    }
}
```

**Comment by JordanMartinez** [comment]
@cerebrosoft I tested out your above code and can reproduce the issue using the following TestFX test:
````java
import javafx.scene.input.KeyCode;
import org.fxmisc.richtext.InlineCssTextAreaAppTest;
import org.fxmisc.wellbehaved.event.EventPattern;
import org.fxmisc.wellbehaved.event.InputMap;
import org.fxmisc.wellbehaved.event.Nodes;
import org.hamcrest.CoreMatchers;
import org.junit.Test;
import org.testfx.api.FxAssert;
import org.testfx.util.DebugUtils;

public class KeyboardOverrides extends InlineCssTextAreaAppTest {

    @Test
    public void testOverrideTabKeyBehavior() {
        String spaces = "    ";
        interact(() -> {
            Nodes.addInputMap(area,
                    InputMap.consume(
                            EventPattern.keyTyped()
                                // factor out possibility of shift/control/alt being pressed
                                .onlyIf(e -> e.getCharacter().equals("\t")),
                            e -> {
                                area.replaceSelection(spaces);
                                // insure event truly is consumed
                                e.consume();
                            }));
        });

        type(KeyCode.TAB);

        String tabCharReplacement = "#";
        String spaceCharReplacment = "!";

        String actual = area.getText()
                .replace("\t", tabCharReplacement)
                .replace(" ", spaceCharReplacment);
        String expected = spaces
                .replace(" ", spaceCharReplacment);

        FxAssert.verifyThat(expected, CoreMatchers.equalTo(actual),
                DebugUtils.informedErrorMessage("Context", false, true, this, true, true));
    }
}
````
This outputs the following error message:
````java
java.lang.AssertionError: 
Expected: "#!!!!"
     but: was "!!!!"

   Context
      Keys pressed at test failure:
      Mouse Buttons pressed at test failure:
      Fired events since test began:
         WindowEvent [source = javafx.stage.Stage@366e2eef, target = javafx.stage.Stage@366e2eef, eventType = WINDOW_SHOWING, consumed = false]
         WindowEvent [source = javafx.stage.Stage@366e2eef, target = javafx.stage.Stage@366e2eef, eventType = WINDOW_SHOWN, consumed = false]
         MouseEvent [source = javafx.stage.Stage@366e2eef, target = javafx.scene.Scene@6df97b55, eventType = MOUSE_ENTERED_TARGET, consumed = false, x = 318.0, y = 64.0, z = 0.0, button = NONE, pickResult = PickResult [node = InlineCssTextArea@7081a20e[styleClass=root styled-text-area], point = Point3D [x = 318.0, y = 64.0, z = 0.0], distance = 746.4101615137755]
         MouseEvent [source = javafx.stage.Stage@366e2eef, target = InlineCssTextArea@7081a20e[styleClass=root styled-text-area], eventType = MOUSE_ENTERED_TARGET, consumed = false, x = 318.0, y = 64.0, z = 0.0, button = NONE, pickResult = PickResult [node = InlineCssTextArea@7081a20e[styleClass=root styled-text-area], point = Point3D [x = 318.0, y = 64.0, z = 0.0], distance = 746.4101615137755]
         MouseEvent [source = javafx.stage.Stage@366e2eef, target = VirtualFlow@6ffe1684[styleClass=virtual-flow], eventType = MOUSE_ENTERED_TARGET, consumed = false, x = 318.0, y = 64.0, z = 0.0, button = NONE, pickResult = PickResult [node = Navigator@42e115e7, point = Point3D [x = 318.0, y = 64.0, z = 0.0], distance = 746.4101615137755]
         MouseEvent [source = javafx.stage.Stage@366e2eef, target = Navigator@42e115e7, eventType = MOUSE_ENTERED_TARGET, consumed = false, x = 318.0, y = 64.0, z = 0.0, button = NONE, pickResult = PickResult [node = Navigator@42e115e7, point = Point3D [x = 318.0, y = 64.0, z = 0.0], distance = 746.4101615137755]
         KeyEvent [source = javafx.stage.Stage@366e2eef, target = InlineCssTextArea@7081a20e[styleClass=root styled-text-area], eventType = KEY_PRESSED, consumed = false, character =  , text = 	, code = TAB]
         KeyEvent [source = javafx.stage.Stage@366e2eef, target = InlineCssTextArea@7081a20e[styleClass=root styled-text-area], eventType = KEY_TYPED, consumed = false, character = 	, text = , code = UNDEFINED]
         KeyEvent [source = javafx.stage.Stage@366e2eef, target = InlineCssTextArea@7081a20e[styleClass=root styled-text-area], eventType = KEY_RELEASED, consumed = false, character =  , text = 	, code = TAB] <Click to see difference>


	at org.testfx.api.FxAssert.verifyThatImpl(FxAssert.java:223)
	at org.testfx.api.FxAssert.verifyThat(FxAssert.java:68)
	at org.fxmisc.richtext.override.KeyboardOverrides.testOverrideTabKeyBehavior(KeyboardOverrides.java:39)
````
I'm honestly baffled why it doesn't work. I initially thought the issue was with WellBehavedFX not consuming the event for some reason. However, then I added the `KeyEvent#consume()` method in the overriding `InputMap`. The output showed that, whether the method is included or not, the event is never consumed. I'm guessing this is a JDK bug? I ran the tests using:
````
java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)
````

**Comment by JordanMartinez** [comment]
@liz3 Since you said my solution worked, which OS are you using?

I'm using Linux Mint 18.3. @cerebrosoft, what's your OS?

**Comment by cerebrosoft** [comment]
Thanks for looking into this- I'm on a Windows 10 box.

**Comment by JordanMartinez** [comment]
Thanks for your response. Let's continue the discussion in #707.

**Comment by JordanMartinez** [comment]
@cerebrosoft I finally figured out what was causing the issue :man_facepalming: 

The default behavior uses 
````java
InputMapTemplate.consume(
   EventPattern.keyPressed(KeyCode.TAB),
   (behavior, keyevent) -> behavior.getView().replaceSelection("\t")
);
````
I mislead you and @liz3 by saying the `EventPattern` needed to be `EventPattern.keyTyped("\t")`. This actually doesn't override the default behavior (keyTyped vs keyPressed). I have since corrected [my original answer](https://github.com/FXMisc/RichTextFX/issues/699#issuecomment-370631560) to her question.
